tiledb/sm/array_schema/dimension.cc

```diff
@@ -399,40 +417,50 @@ Status Dimension::check_tile_extent() const {
 
   auto tile_extent = static_cast<T*>(tile_extent_);
   auto domain = static_cast<T*>(domain_);
+  bool is_int = std::is_integral<T>::value;
 
   // Check if tile extent is negative or 0
   if (*tile_extent <= 0)
     return LOG_STATUS(Status::DimensionError(
         "Tile extent check failed; Tile extent must be greater than 0"));
 
   // Check if tile extent exceeds domain
-  if (&typeid(T) == &typeid(float) || &typeid(T) == &typeid(double)) {
+  if (!is_int) {
     if (*tile_extent > domain[1] - domain[0])
       return LOG_STATUS(
           Status::DimensionError("Tile extent check failed; Tile extent "
                                  "exceeds dimension domain range"));
   } else {
     // Check if tile extent exceeds domain
-    // TODO (sp): Handle negative domains
-    if ((domain[1] - domain[0] < std::numeric_limits<T>::max() - 1) &&
-        (*tile_extent > domain[1] - domain[0] + 1))
+    uint64_t range = domain[1] - domain[0] + 1;
+    if (uint64_t(*tile_extent) > range)
       return LOG_STATUS(
           Status::DimensionError("Tile extent check failed; Tile extent "
                                  "exceeds dimension domain range"));
-  }
 
-  // In the worst case one tile_extent will be added to the upper domain,
-  // so check if the expanded domain will exceed type T's max limit.
-  auto upper_floor = (domain[1] / (*tile_extent)) * (*tile_extent);
-  if (upper_floor != domain[1] &&
-      upper_floor > std::numeric_limits<T>::max() - (*tile_extent))
-    return LOG_STATUS(Status::DimensionError(
-        "Tile extent check failed; domain max expanded to multiple of tile "
-        "extent exceeds max value representable by domain type. Reduce domain "
-        "max by 1 tile extent to allow for expansion."));
+    // In the worst case one tile extent will be added to the upper domain
+    // for the dense case, so check if the expanded domain will exceed type
+    // T's max limit.
+    if (range % uint64_t(*tile_extent)) {
+      auto upper_floor =
+          ((range - 1) / (*tile_extent)) * (*tile_extent) + domain[0];
```

Comment:
I am suspicious of this computation w.r.t. negative domains, especially with `auto` type. Suppose domain is `[-100, -87]` with tile extent 5. Then `range == 13`, and `upper_floor == 12/5 * 5 - 100 == 10 - 100 == -90`. Does this cause problems if `upper_floor` is negative?

Reply:
This is correct. Essentially later we check whether `-90` (`upper_floor`) plus `4` (tile extent minus one) exceeds the domain type max value.